File: forms/united-kingdom-nhs-england-medical-exemption-certificate/front-end-form-with-svelte/src/lib/engine/fp92a-validator.test.ts
import { describe, expect, it } from 'vitest';
import { evaluateFp92a } from './fp92a-validator';
import { fp92aRules } from './fp92a-rules';
import { detectAdditionalFlags } from './flagged-issues';
import type {
	EligibleConditionCode,
	Fp92aApplication,
	QualifyingConditionDetail
} from './types';

const ALL_CODES: EligibleConditionCode[] = [
	'permanent-fistula',
	'hypoadrenalism',
	'diabetes-insipidus-or-hypopituitarism',
	'diabetes-mellitus-not-diet-only',
	'hypoparathyroidism',
	'myasthenia-gravis',
	'myxoedema',
	'epilepsy-on-anticonvulsant',
	'continuing-physical-disability',
	'cancer-or-effects'
];

function defaultCondition(code: EligibleConditionCode): QualifyingConditionDetail {
	return {
		code,
		selected: false,
		diagnosisDate: '',
		snomedCtCode: '',
		icd10Code: '',
		treatmentDetail: '',
		fistulaSite: '',
		applianceType: '',
		substitutionTherapy: '',
		onSubstitutionTherapy: '',
		diabetesTreatmentMode: '',
		anticonvulsant: '',
		continuousAnticonvulsantTherapy: '',
		cannotLeaveHomeUnaided: '',
		disabilityCarerDetail: '',
		disabilityExpectedToBePermanent: '',
		cancerSite: '',
		cancerTreatmentPhase: '',
		histologyConfirmed: '',
		practitionerAttestationNotes: ''
	};
}

function emptyApplication(): Fp92aApplication {
	return {
		practitioner: {
			name: '',
			role: '',
			registrationBody: '',
			registrationNumber: '',
			practiceName: '',
			practiceCode: '',
			practitionerCode: '',
			postalAddressAsFullText: '',
			postcode: '',
			countryAsIso31661Alpha2: '',
			phone: '',
			email: '',
			completionDate: ''
		},
		patient: {
			title: '',
			surname: '',
			forenames: '',
			name: '',
			birthDate: '',
			sex: '',
			postalAddressAsFullText: '',
			postcode: '',
			countryAsIso31661Alpha2: '',
			unitedKingdomNhsNumber: '',
			phone: '',
			email: '',
			fullTimeEducation: '',
			pregnancyStatus: ''
		},
		existingExemption: {
			hasExistingCertificate: '',
			applicationKind: '',
			previousCertificateNumber: '',
			previousCertificateExpiryDate: ''
		},
		ageCheck: { practitionerAcknowledgedAgeAdvice: '' },
		pregnancyCheck: { practitionerAcknowledgedFw8Redirect: '' },
		qualifyingConditions: ALL_CODES.map(defaultCondition),
		declaration: {
			practitionerSignaturePresent: '',
			practitionerHasAccessToMedicalRecords: '',
			practitionerDeclarationText: '',
			signatureDate: ''
		}
	};
}

/** Fully-populated application with diabetes mellitus (insulin) — eligible. */
function eligibleDiabetesApplication(): Fp92aApplication {
	const d = emptyApplication();
	d.practitioner.name = 'Dr Smith';
	d.practitioner.role = 'gp';
	d.practitioner.registrationBody = 'GMC';
	d.practitioner.registrationNumber = '1234567';
	d.practitioner.practiceName = 'High Street Surgery';
	d.practitioner.completionDate = '2026-05-18';

	d.patient.surname = 'Doe';
	d.patient.forenames = 'Jane';
	d.patient.birthDate = '1980-05-10';
	d.patient.postalAddressAsFullText = '1 High Street, London';
	d.patient.postcode = 'SW1A 1AA';
	d.patient.unitedKingdomNhsNumber = '9990123456';
	d.patient.pregnancyStatus = 'not-pregnant';

	const c = d.qualifyingConditions.find(
		(x) => x.code === 'diabetes-mellitus-not-diet-only'
	);
	if (c) {
		c.selected = true;
		c.diagnosisDate = '2010-01-01';
		c.diabetesTreatmentMode = 'insulin';
		c.treatmentDetail = 'Basal-bolus insulin';
	}

	d.declaration.practitionerSignaturePresent = 'yes';
	d.declaration.practitionerHasAccessToMedicalRecords = 'yes';
	d.declaration.signatureDate = '2026-05-18';
	return d;
}

describe('FP92A — rule set', () => {
	it('all rule IDs are unique', () => {
		const ids = fp92aRules.map((r) => r.id);
		expect(new Set(ids).size).toBe(ids.length);
	});

	it('every rule has a non-empty message', () => {
		for (const r of fp92aRules) {
			expect(r.message.length).toBeGreaterThan(0);
		}
	});
});

describe('FP92A — empty application', () => {
	it('is ineligible and fires the urgent completeness rules', () => {
		const r = evaluateFp92a(emptyApplication());
		expect(r.outcome).toBe('ineligible');
		const ids = r.firedRules.map((f) => f.id);
		expect(ids).toContain('FP92A-COMP-CONDITION-SELECTED');
		expect(ids).toContain('FP92A-COMP-SIGNATURE');
		expect(ids).toContain('FP92A-COMP-NHS-NUMBER');
	});

	it('sorts fired rules urgent → low', () => {
		const r = evaluateFp92a(emptyApplication());
		const order = { urgent: 0, high: 1, medium: 2, low: 3 } as const;
		const priorities = r.firedRules.map((f) => f.priority);
		const sorted = [...priorities].sort((a, b) => order[a] - order[b]);
		expect(priorities).toEqual(sorted);
	});
});

describe('FP92A — eligible diabetes application', () => {
	it('is eligible with diabetes-mellitus-not-diet-only', () => {
		const r = evaluateFp92a(eligibleDiabetesApplication());
		expect(r.outcome).toBe('eligible');
		expect(r.eligibleConditions).toContain('diabetes-mellitus-not-diet-only');
		expect(r.validFrom).toBe('2026-05-18');
		expect(r.validUntil).toBe('2031-05-18');
		expect(r.redirectTo).toBe('');
	});

	it('is ineligible when diabetes is diet-only', () => {
		const d = eligibleDiabetesApplication();
		const c = d.qualifyingConditions.find(
			(x) => x.code === 'diabetes-mellitus-not-diet-only'
		);
		if (c) c.diabetesTreatmentMode = 'diet-only';
		const r = evaluateFp92a(d);
		expect(r.outcome).toBe('ineligible');
		const ids = r.firedRules.map((f) => f.id);
		expect(ids).toContain('FP92A-DISQ-DIET-ONLY-DIABETES');
	});
});

describe('FP92A — redirects', () => {
	it('redirects pregnant applicants to FW8', () => {
		const d = eligibleDiabetesApplication();
		d.patient.pregnancyStatus = 'pregnant';
		const r = evaluateFp92a(d);
		expect(r.redirectTo).toBe('FW8');
		expect(r.outcome).toBe('ineligible');
	});

	it('redirects 60+ applicants to age exemption', () => {
		const d = eligibleDiabetesApplication();
		d.patient.birthDate = '1950-01-01';
		const r = evaluateFp92a(d);
		expect(r.redirectTo).toBe('age-exemption');
		const ids = r.firedRules.map((f) => f.id);
		expect(ids).toContain('FP92A-REDIRECT-AGE-60-PLUS');
	});

	it('redirects under-16 applicants to age exemption', () => {
		const d = eligibleDiabetesApplication();
		d.patient.birthDate = '2015-01-01';
		const r = evaluateFp92a(d);
		expect(r.redirectTo).toBe('age-exemption');
	});

	it('redirects 16-18 in education to age exemption', () => {
		const d = eligibleDiabetesApplication();
		d.patient.birthDate = '2009-01-01';
		d.patient.fullTimeEducation = 'yes';
		const r = evaluateFp92a(d);
		expect(r.redirectTo).toBe('age-exemption');
	});
});

describe('FP92A — requires-clarification', () => {
	it('requires clarification when cancer histology is pending', () => {
		const d = eligibleDiabetesApplication();
		const diab = d.qualifyingConditions.find(
			(x) => x.code === 'diabetes-mellitus-not-diet-only'
		);
		if (diab) diab.selected = false;
		const cancer = d.qualifyingConditions.find((x) => x.code === 'cancer-or-effects');
		if (cancer) {
			cancer.selected = true;
			cancer.cancerSite = 'breast';
			cancer.cancerTreatmentPhase = 'active-treatment';
			cancer.histologyConfirmed = 'pending';
		}
		const r = evaluateFp92a(d);
		expect(r.outcome).toBe('requires-clarification');
	});
});

describe('FP92A — disqualifying / disability', () => {
	it('is ineligible when disability is declared temporary', () => {
		const d = eligibleDiabetesApplication();
		const diab = d.qualifyingConditions.find(
			(x) => x.code === 'diabetes-mellitus-not-diet-only'
		);
		if (diab) diab.selected = false;
		const dis = d.qualifyingConditions.find(
			(x) => x.code === 'continuing-physical-disability'
		);
		if (dis) {
			dis.selected = true;
			dis.cannotLeaveHomeUnaided = 'yes';
			dis.disabilityExpectedToBePermanent = 'no';
		}
		const r = evaluateFp92a(d);
		expect(r.outcome).toBe('ineligible');
		const ids = r.firedRules.map((f) => f.id);
		expect(ids).toContain('FP92A-DISQ-TEMP-DISABILITY');
	});
});

describe('FP92A — additional flags', () => {
	it('flags missing NHS number as urgent', () => {
		const d = eligibleDiabetesApplication();
		d.patient.unitedKingdomNhsNumber = '';
		const flags = detectAdditionalFlags(d);
		expect(flags.some((f) => f.id === 'FLAG-MISSING-NHS-NUMBER' && f.priority === 'urgent')).toBe(
			true
		);
	});

	it('flags missing practitioner signature as urgent', () => {
		const d = eligibleDiabetesApplication();
		d.declaration.practitionerSignaturePresent = '';
		const flags = detectAdditionalFlags(d);
		expect(
			flags.some((f) => f.id === 'FLAG-MISSING-SIGNATURE' && f.priority === 'urgent')
		).toBe(true);
	});

	it('flags pregnancy as high', () => {
		const d = eligibleDiabetesApplication();
		d.patient.pregnancyStatus = 'pregnant';
		const flags = detectAdditionalFlags(d);
		expect(flags.some((f) => f.id === 'FLAG-PREGNANCY-FW8' && f.priority === 'high')).toBe(true);
	});

	it('flags age-60+ as medium', () => {
		const d = eligibleDiabetesApplication();
		d.patient.birthDate = '1950-01-01';
		const flags = detectAdditionalFlags(d);
		expect(
			flags.some((f) => f.id === 'FLAG-AGE-EXEMPTION-60-PLUS' && f.priority === 'medium')
		).toBe(true);
	});
});

File: forms/united-kingdom-nhs-england-medical-exemption-certificate/full-stack-with-loco-tera-htmx-alpine/tests/engine/fp92a_validator_test.rs
use united_kingdom_nhs_england_medical_exemption_certificate_tera_crate::engine::fp92a_rules::fp92a_rules;
use united_kingdom_nhs_england_medical_exemption_certificate_tera_crate::engine::fp92a_validator::validate_fp92a;
use united_kingdom_nhs_england_medical_exemption_certificate_tera_crate::engine::types::*;

fn fully_populated_eligible() -> ApplicationData {
    let mut d = ApplicationData::default();
    d.patient.surname = "Smith".into();
    d.patient.forenames = "Jane".into();
    d.patient.birth_date = "1980-04-01".into();
    d.patient.united_kingdom_nhs_number = "9434765919".into();
    d.patient.postal_address_as_full_text = "1 High St, Newcastle".into();
    d.patient.pregnancy_status = "not-pregnant".into();
    d.practitioner.name = "Dr Alex Doctor".into();
    d.practitioner.registration_body = "GMC".into();
    d.practitioner.registration_number = "1234567".into();
    d.practitioner.practice_name = "High St Surgery".into();
    d.practitioner.completed_date = "2026-05-18".into();
    d.declaration.signature_present = "yes".into();
    d.declaration.access_to_medical_records = "yes".into();
    d.conditions.push(QualifyingCondition {
        code: "epilepsy-on-anticonvulsant".into(),
        selected: "yes".into(),
        continuous_anticonvulsant_therapy: "yes".into(),
        anticonvulsant: "Levetiracetam 500mg BD".into(),
        ..Default::default()
    });
    d
}

#[test]
fn empty_application_is_ineligible() {
    let r = validate_fp92a(&ApplicationData::default());
    assert_eq!(r.outcome, "ineligible");
    assert!(r
        .fired_rules
        .iter()
        .any(|x| x.id == "FP92A-CP-NO-CONDITION-SELECTED"));
}

#[test]
fn pregnancy_redirects_to_fw8() {
    let mut d = fully_populated_eligible();
    d.patient.pregnancy_status = "pregnant".into();
    let r = validate_fp92a(&d);
    assert_eq!(r.redirect_to, "FW8");
    assert_eq!(r.outcome, "ineligible");
}

#[test]
fn diet_only_diabetes_is_disqualifying() {
    let mut d = fully_populated_eligible();
    d.conditions.clear();
    d.conditions.push(QualifyingCondition {
        code: "diabetes-mellitus-not-diet-only".into(),
        selected: "yes".into(),
        diabetes_treatment_mode: "diet-only".into(),
        ..Default::default()
    });
    let r = validate_fp92a(&d);
    assert!(r
        .fired_rules
        .iter()
        .any(|x| x.id == "FP92A-DQ-DIET-ONLY-DIABETES"));
    assert_ne!(r.outcome, "eligible");
}

#[test]
fn eligible_when_epilepsy_attested_and_all_fields_complete() {
    let d = fully_populated_eligible();
    let r = validate_fp92a(&d);
    assert_eq!(r.outcome, "eligible", "fired: {:?}", r.fired_rules);
    assert!(r
        .eligible_condition_codes
        .contains(&"epilepsy-on-anticonvulsant".to_string()));
}

#[test]
fn cancer_pending_histology_requires_clarification() {
    let mut d = fully_populated_eligible();
    d.conditions.clear();
    d.conditions.push(QualifyingCondition {
        code: "cancer-or-effects".into(),
        selected: "yes".into(),
        cancer_site: "Breast".into(),
        cancer_treatment_phase: "active-treatment".into(),
        histology_confirmed: "pending".into(),
        ..Default::default()
    });
    let r = validate_fp92a(&d);
    assert_eq!(r.outcome, "requires-clarification");
}

#[test]
fn rule_ids_are_unique() {
    let rules = fp92a_rules();
    let mut ids: Vec<&str> = rules.iter().map(|r| r.id).collect();
    ids.sort();
    let len_before = ids.len();
    ids.dedup();
    assert_eq!(ids.len(), len_before, "rule IDs must be unique");
}

#[test]
fn validity_window_is_five_years() {
    let d = fully_populated_eligible();
    let r = validate_fp92a(&d);
    assert_eq!(r.validity_years, 5);
    assert!(!r.valid_from.is_empty());
    assert!(!r.valid_until.is_empty());
}
